// @near-js/accounts/lib/commonjs/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  Account: () => import_account.Account,
  AccountCreator: () => import_account_creator.AccountCreator,
  ArgumentSchemaError: () => import_errors.ArgumentSchemaError,
  ConflictingOptions: () => import_errors.ConflictingOptions,
  Connection: () => import_connection.Connection,
  Contract: () => import_contract.Contract,
  LocalAccountCreator: () => import_account_creator.LocalAccountCreator,
  MULTISIG_ALLOWANCE: () => import_constants.MULTISIG_ALLOWANCE,
  MULTISIG_CHANGE_METHODS: () => import_constants.MULTISIG_CHANGE_METHODS,
  MULTISIG_CONFIRM_METHODS: () => import_constants.MULTISIG_CONFIRM_METHODS,
  MULTISIG_DEPOSIT: () => import_constants.MULTISIG_DEPOSIT,
  MULTISIG_GAS: () => import_constants.MULTISIG_GAS,
  MULTISIG_STORAGE_KEY: () => import_constants.MULTISIG_STORAGE_KEY,
  MultisigDeleteRequestRejectionError: () => import_types.MultisigDeleteRequestRejectionError,
  MultisigStateStatus: () => import_types.MultisigStateStatus,
  UnknownArgumentError: () => import_errors.UnknownArgumentError,
  UnsupportedSerializationError: () => import_errors.UnsupportedSerializationError,
  UrlAccountCreator: () => import_account_creator.UrlAccountCreator
});
module.exports = __toCommonJS(index_exports);
var import_account = require('./account.cjs');
var import_account_creator = require('./account_creator.cjs');
var import_connection = require('./connection.cjs');
var import_constants = require('./constants.cjs');
var import_contract = require('./contract.cjs');
var import_errors = require('./errors.cjs');
var import_types = require('./types.cjs');
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Account,
  AccountCreator,
  ArgumentSchemaError,
  ConflictingOptions,
  Connection,
  Contract,
  LocalAccountCreator,
  MULTISIG_ALLOWANCE,
  MULTISIG_CHANGE_METHODS,
  MULTISIG_CONFIRM_METHODS,
  MULTISIG_DEPOSIT,
  MULTISIG_GAS,
  MULTISIG_STORAGE_KEY,
  MultisigDeleteRequestRejectionError,
  MultisigStateStatus,
  UnknownArgumentError,
  UnsupportedSerializationError,
  UrlAccountCreator
});
